Refactor importScripts(urls)

jungkees opened the pull request:
Refactor importScripts(urls) by extracting it into an algorithm,  import scripts
into worker global scope, having hooks that can be overriden by derived
interfaces. The service worker spec uses the extracted algorithm with the
overriden options: https://slightlyoff.github.io/ServiceWorker/spec/service_worker/#importscripts

Fixes https://www.w3.org/Bugs/Public/show_bug.cgi?id=28737. Please take a look!

domenic commented:
Thanks! This is a good start. Let's structure this slightly differently however. I'm thinking something like:

> When a script invokes the importScripts(urls) method on a WorkerGlobalScope object, the user agent must import scripts into worker global scope, with the following options:
>
> - To validate the state, ...
> - To get a fetch result, ...
> - To postprocess the fetch result, ...

(notice my slight renaming of the hooks)

Then, after this, similar to what you've already got, you say:

> The import scripts into worker global scope algorithm has three hooks...

> To import scripts into worker global scope, the user agent must run the following steps:...

---

Also, not a big deal but try to follow the commit message guidelines from [writing good commit messages](https://github.com/erlang/otp/wiki/Writing-good-commit-messages).

jungkees commented:
Thanks for the review and the pointer! I updated the snapshot and changed the title and the description in the OP as well as the commit message. Once it's rolled, I'll update the relevant part of the service worker spec.

annevk commented:
@jungkees can you rebase your branch and force push it as a single commit?

jungkees commented:
@annevk I just uploaded a new snapshot as such. Please review.

annevk commented:
My other minor nits would be that I would prefer "Fixes: ..." over "BUG=..." in the commit message and that I would drop "index.html" from the service worker link in the commit message, but I'll leave that up to you since it's not super important.

Will assign to @domenic for final review.

Thanks a lot for doing this @jungkees!

jungkees commented:
Thanks for the review. I'm happy to do this and remove the fixme label for the monkey patch in the service worker spec. :)

domenic commented:
Awesome! I made a few minor tweaks and merged as https://github.com/whatwg/html/commit/509578d97bd12374e50a409bd4f1ec63cd36a543. It makes me really happy that we can do things like this now!

jungkees commented:
Great! SW addressed this change: 59e3745a04239197fb3177c329f1e5f4afeb885d.
